packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs

```diff
@@ -959,7 +934,126 @@ where
+
+    async fn lock_entity_for_update(&self, entity_id: EntityId) -> Result<(), QueryError> {
+        self.as_client()
+            .query_one(
+                // TODO: consider if this row locking is problematic with Citus.
+                //   `FOR UPDATE` is only allowed in single-shard queries.
+                //   https://docs.citusdata.com/en/stable/develop/reference_workarounds.html#sql-support-and-workarounds
+                //   see: https://app.asana.com/0/0/1203284257408542/f
+                r#"
+                SELECT * FROM entities
+                WHERE entity_id = $1
+                FOR UPDATE;
+                "#,
+                &[&entity_id],
+            )
+            .await
+            .into_report()
+            .change_context(QueryError)?;
+
+        Ok(())
+    }
+
+    async fn move_entity_to_histories(
+        &self,
+        entity_id: EntityId,
+        archived: bool,
+    ) -> Result<PersistedEntityMetadata, InsertionError> {
+        let historic_entity = self
+            .as_client()
+            .query_one(
+                r#"
+                -- First we delete the _latest_ entity from the entities table.
+                WITH to_move_to_historic AS (
+                    DELETE FROM entities
+                    WHERE entity_id = $1
+                    RETURNING
+                        entity_id, version,
+                        entity_type_version_id,
+                        properties,
+                        left_order, right_order,
+                        left_entity_id, right_entity_id,
+                        owned_by_id, created_by_id, updated_by_id
+                ),
+                inserted_in_historic AS (
+                    -- We immediately put this deleted entity into the historic table.
+                    -- As this should be done in a transaction, we should be safe that this move
+                    -- doesn't produce invalid state.
+                    INSERT INTO entity_histories(
+                        entity_id, version,
+                        entity_type_version_id,
+                        properties,
+                        left_order, right_order,
+                        left_entity_id, right_entity_id,
+                        owned_by_id, created_by_id, updated_by_id,
+                        archived
+                    )
+                    SELECT
+                        entity_id, version,
+                        entity_type_version_id,
+                        properties,
+                        left_order, right_order,
+                        left_entity_id, right_entity_id,
+                        owned_by_id, created_by_id, updated_by_id,
+                        $2::boolean
+                    FROM to_move_to_historic
+                    -- We only return metadata
+                    RETURNING
+                        entity_id, version,
+                        entity_type_version_id,
+                        left_order, right_order,
```

Comment:
This is making me think we want to rework the metadata structs even further. Probably do not want to be returning the `order` and all of that. (But we can revisit this separately to the PR)